pkg/distributors/contracts/DistributionScheduler.sol

```diff
@@ -0,0 +1,142 @@
+
+    constructor(IMultiDistributor multiDistributor) {
+        _multiDistributor = multiDistributor;
+    }
+
+    enum DistributionStatus { UNINITIALIZED, PENDING, STARTED }
+
+    struct ScheduledDistribution {
+        bytes32 distributionId;
+        IERC20 stakingToken;
+        IERC20 distributionToken;
+        uint256 startTime;
+        address owner;
+        uint256 amount;
+        DistributionStatus status;
+    }
+
+    event DistributionScheduled(
+        bytes32 scheduleId,
+        address indexed owner,
+        IERC20 indexed stakingToken,
+        IERC20 indexed distributionToken,
+        uint256 startTime,
+        uint256 amount
+    );
+    event DistributionStarted(
+        bytes32 scheduleId,
+        address indexed owner,
+        IERC20 indexed stakingToken,
+        IERC20 indexed distributionToken,
+        uint256 startTime,
+        uint256 amount
+    );
+
+    mapping(bytes32 => ScheduledDistribution) private _scheduledDistributions;
+
+    function getScheduledDistributionInfo(bytes32 scheduleId)
+        external
+        view
+        returns (ScheduledDistribution memory reward)
+    {
+        return _scheduledDistributions[scheduleId];
+    }
+
+    function startDistributions(bytes32[] calldata scheduleIds) external {
+        for (uint256 i; i < scheduleIds.length; i++) {
+            bytes32 scheduleId = scheduleIds[i];
+            ScheduledDistribution memory scheduledDistribution = _scheduledDistributions[scheduleId];
+
+            require(scheduledDistribution.status == DistributionStatus.PENDING, "Distribution cannot be started");
+            require(scheduledDistribution.startTime <= block.timestamp, "Distribution start time is in the future");
+
+            _scheduledDistributions[scheduleId].status = DistributionStatus.STARTED;
+
+            uint256 allowance = scheduledDistribution.distributionToken.allowance(
+                address(this),
+                address(_multiDistributor)
+            );
+            if (allowance < scheduledDistribution.amount) {
+                scheduledDistribution.distributionToken.approve(address(_multiDistributor), type(uint256).max);
+            }
+            _multiDistributor.fundDistribution(scheduledDistribution.distributionId, scheduledDistribution.amount);
+            emit DistributionStarted(
+                scheduleId,
+                scheduledDistribution.owner,
+                scheduledDistribution.stakingToken,
+                scheduledDistribution.distributionToken,
+                scheduledDistribution.startTime,
+                scheduledDistribution.amount
+            );
+        }
+    }
+
+    function getScheduleId(
+        IERC20 stakingToken,
+        IERC20 distributionToken,
+        address owner,
+        uint256 startTime
+    ) public pure returns (bytes32) {
+        return keccak256(abi.encodePacked(stakingToken, distributionToken, owner, startTime));
+    }
+
+    function scheduleDistribution(
```

Comment:
This seems incomplete: shouldn't we check that `msg.sender` is the owner of the distro?

Reply:
This is correct. This preexisting issue is going to be fixed in a later PR which makes the scheduler pull a lot more data from the distributor (e.g. it currently has no guarantees that the provided tokens are relevant for the specified distribution)

Reply:
Ah I see, thanks.